src/workerd/api/memory-cache.c++

```diff
@@ -361,9 +362,15 @@ SharedMemoryCache::Use::FallbackDoneCallback SharedMemoryCache::Use::prepareFall
   auto deferredCancel = kj::defer([this, status = kj::mv(status), &inProgress]() {
     // If the callback was destroyed without having run (for example, because
     // it was added to an I/O context that has since been canceled), we treat
-    // it as if the promise had failed.
+    // it as if the promise had failed - but only if handleFailureOnDestruction is true.
     if (!status->hasSettled) {
-      handleFallbackFailure(inProgress);
+      // Use kj::evalLater to break recursion when many waiters are destroyed simultaneously
+      auto promise = kj::evalLater([this, &inProgress]() { handleFallbackQueue(inProgress); });
```

Comment:
I'm worried that, here and in the other `kj::evalLater()`, we're capturing `this`, but we don't actually have any guarantee that the promise we're creating will not outlive `this` (which is inside a JSG object in some isolate, I think), since we're detaching the promise.

It's likely that it will almost always not outlive `this`. But I'm worried about the "almost" part of that sentence. :/

I don't really have any better idea at the moment. Will need to get my head more into this code.